optimum/exporters/openvino/model_patcher.py

```diff
@@ -6456,3 +6456,378 @@ def __exit__(self, exc_type, exc_value, traceback):
             if layer.is_moe_layer:
                 layer.feed_forward.forward = layer.feed_forward._orig_forward
             layer.self_attn.forward = layer.self_attn._orig_forward
+
+
+class ConvSequenceTransform(torch.nn.Module):
+    def __init__(self, conv_kernel_size, use_conv_bias, conv1, act, conv_bias):
+        super().__init__()
+        self.conv_kernel_size = conv_kernel_size
+        self.use_conv_bias = use_conv_bias
+        self.conv1d = conv1
+        self.act = act
+        self.conv_bias = conv_bias
+
+    def update_conv_state(
+        self, conv_state, new_conv_state: torch.Tensor, cache_position: torch.LongTensor
+    ) -> torch.Tensor:
+        conv_state_1 = conv_state.roll(shifts=-1, dims=-1)
+        upd_conv_state = conv_state_1.scatter(2, cache_position, new_conv_state)
+        return upd_conv_state
+
+    def get_positions(self, conv_state, cache_position):
+        cache_position_clamped = cache_position.clamp(0, self.conv_kernel_size - 1)
+        positions = cache_position_clamped.expand(conv_state.shape[0], conv_state.shape[1], -1)
+        return positions
+
+    def forward(self, hidden_states, cache_position, conv_state):
+        pad_value = (self.conv_kernel_size - hidden_states.shape[-1]) * (
+            cache_position.shape[0] == self.conv_kernel_size
+        )
+        new_conv_state = torch.nn.functional.pad(hidden_states, (pad_value, 0))
+        upd_cache_position = self.get_positions(conv_state, cache_position)
+        upd_conv_state = self.update_conv_state(conv_state, new_conv_state, upd_cache_position)
+        if cache_position.shape[0] == self.conv_kernel_size:
+            hidden_states = self.conv1d(hidden_states)[:, :, : hidden_states.shape[-1]]
+        else:
+            hidden_states = torch.sum(upd_conv_state * self.conv1d.weight[:, 0, :], dim=-1)
+            hidden_states += self.conv_bias
+            hidden_states = hidden_states.unsqueeze(-1)
+        hidden_states = self.act(hidden_states)  # [batch, intermediate_size, seq_len]
+        return hidden_states, upd_conv_state
+
+
+class SelectiveScan(torch.nn.Module):
+    def forward(self, ssm, u, dt, A, B, C, D):
+        dA = torch.einsum("bld,dn->bldn", dt, A)
+        dB_u = torch.einsum("bld,bld,bln->bldn", dt, u, B)
+        dA_cumsum = torch.nn.functional.pad(dA[:, 1:], (0, 0, 0, 0, 0, 1)).flip(1).cumsum(1).exp().flip(1)
+        x = dB_u * dA_cumsum + (ssm.unsqueeze(1) * dA[:, :1].exp())
+        x = x.cumsum(1) / (dA_cumsum + 1e-12)
+        y = torch.einsum("bldn,bln->bld", x, C)
+        return y + u * D, x[:, -1, :, :]
+
+
+def mamba_mixer_forward(
```

Comment:
would you mind adding a link to the original code as well as highlight which part was modified and why it's needed)
https://github.com/huggingface/transformers/blob/v4.53.0/src/transformers/models/mamba/modeling_mamba.py#L233

Reply:
Done